src/cpp/session/modules/SessionRParser.cpp

```diff
@@ -0,0 +1,1287 @@
+         return;                                                               \
+      }                                                                        \
+   } while (0)
+
+
+#define UNEXPECTED_TOKEN(__CURSOR__, __STATUS__)                               \
+   do                                                                          \
+   {                                                                           \
+      DEBUG("(" << __LINE__ << "): Unexpected token (" << __CURSOR__ << ")");  \
+      __STATUS__.lint().unexpectedToken(__CURSOR__);                           \
+   } while (0)
+
+void lookAheadAndWarnOnUsagesOfSymbol(const RTokenCursor& startCursor,
+                                      RTokenCursor& clone,
+                                      ParseStatus& status)
+{
+   std::size_t braceBalance = 0;
+   std::wstring symbol = startCursor.content();
+   
+   do
+   {
+      // Skip over 'for' arg-list
+      if (clone.contentEquals(L"for"))
+      {
+         if (!clone.moveToNextSignificantToken())
+            return;
+         
+         if (!clone.fwdToMatchingToken())
+            return;
+         
+         continue;
+      }
+      
+      // Skip over functions
+      if (clone.contentEquals(L"function"))
+      {
+         if (!clone.moveToNextSignificantToken())
+            return;
+         
+         if (!clone.fwdToMatchingToken())
+            return;
+         
+         if (!clone.moveToNextSignificantToken())
+            return;
+         
+         if (clone.isType(RToken::LBRACE))
+         {
+            if (!clone.fwdToMatchingToken())
+               return;
+         }
+         else
+         {
+            if (!clone.moveToEndOfStatement(status.isInParentheticalScope()))
+               return;
+         }
+         
+         if (clone.isAtEndOfExpression())
+            return;
+         
+         continue;
+      }
+      
+      const RToken& prev = clone.previousSignificantToken();
+      const RToken& next = clone.nextSignificantToken();
+      
+      if (clone.contentEquals(symbol) &&
+          !isRightAssign(prev) &&
+          !isLeftAssign(next) &&
+          !isLeftBracket(next) &&
+          !isExtractionOperator(prev) &&
+          !isExtractionOperator(next))
+      {
+         status.lint().noSymbolNamed(clone);
+      }
+      
+      braceBalance += isLeftBracket(clone);
+      if (isRightBracket(clone))
+      {
+         --braceBalance;
+         if (isBinaryOp(clone.nextSignificantToken()))
+         {
+            if (!clone.moveToNextSignificantToken())
+               return;
+            continue;
+         }
+         
+         if (braceBalance <= 0)
+            return;
+      }
+      
+      if (clone.nextToken().contentContains(L'\n') ||
+          clone.isType(RToken::SEMI))
+      {
+         return;
+      }
+      
+   } while (clone.moveToNextSignificantToken());
+}
+
+void handleIdentifier(RTokenCursor& cursor,
```

Comment:
We typically use raw pointers for mutable parameters (so it's explicit both at the call site and in the function that a parameter is being mutated)